### DISCUSSION TITLE
Two instances of app possible?

### DISCUSSION THREAD

**Discussion by chazhenry** [airsdk/Adobe-Runtime-Support#1305] [Ideas]
I'd like the user to be able to run two (or more) instances of my app. Currently, if you call the app again from the Windows menu, it just brings the running app to focus. There are some old AS3 tricks to do this, but they don't seem to work with captive runtime. Is it possible?

**Comment by MalacTheLittle** [comment]
You can make copy of the app and just change application id in application descriptor xml (META-INF/AIR/application.xml).

**Reply by chazhenry** [reply]
Are you saying install it twice under different IDs? Or you suggesting code can do this at run time?

**Reply by MalacTheLittle** [reply]
There's post about this [here](https://stackoverflow.com/questions/2217307/starting-an-adobe-air-application-multiple-times). You can make change of application descriptor from runtime, but only if app is not placed in some system protected folder like Program Files.
But you must be carefull with this approach to avoid getting in the loop. You should check apps id and make two different paths like:
if app id is 0, change it to 1 and execute
if app id is 1, just change it to 0

Or, you can install it twice with different id's :)

**Comment by 42retfa** [comment]
Another possible approach :

The main application is constructed as follows :

```
<s:Application>
    <content:ApplicationContent/>
</s:Application>
```

and each time you need a new instance of your application you create a new 

```
<s:Window>
    <content:ApplicationContent/>
</s:Window>
```

**Reply by MalacTheLittle** [reply]
Interesting idea... but wouldn't it ruin performance?

**Reply by 42retfa** [reply]
Performance could be an issue for sure. 
It depends on the content of your app, but it should be manageable given the fact that the user can only interact with one instance at a time.

Vincent.

**Comment by 2jfw** [comment]
@ajwfrost: Though the use case/necessity is probably quite low, I could imagine to have a value in the app-descriptor.xml which determines if an app can be started multiple times or not. I do actually see no reason to limit AIR apps from starting several times - not sure why Adobe did hard-code this behavior (?)

**Reply by marchbold** [reply]
Yeah we've definitely had clients in the past ask for this functionality so would be a good one to make available.

**Comment by ajwfrost** [comment]
I believe the original reasons for this behaviour were (a) some of the features of AIR use the app descriptor/name as a reference value e.g. for LocalConnection functionality or for shared objects etc; (b) that's how macOS works anyway, unless perhaps you launch via 'open -n'...

We had discussed it with Adobe quite some time ago though, when this first came up (when it transpired that people were adding digits to the end of the application ID in order to get around this restriction, but that then caused a problem for our license file checking!) - there's no particular reason why we can't remove this restriction. We had left it in there because we aren't sure what the side-effects might be for some of these other functionalities. But as a starting point, perhaps we can just enable this when it's been explicitly requested by the developer i.e. via an app descriptor flag, and then if you use that you would need to just check that your apps do actually work okay with multiple instances.

This then would follow the general practice of trying not to change anything for apps that don't explicitly want to change, but would mean you can make the deliberate choice to allow it if you wanted to...

thanks

**Reply by ajwfrost** [reply]
Just to update this topic ... we had initially enabled this for command-line apps that are launched via ADL (I thought `apm` was using this capability, or maybe it was a request from someone else making AIR-based command-line tools....)

We've just checked at least on Windows and there's no obvious harmful effects if we allow that switch also for normal windowed apps, at least from an AIR perspective: applications built on AIR may need to consider other implications if there are multiple instances running, as things like the app-storage folder and ELS data will be shared between all the running instances.

So for our forthcoming major update we can add in this capability (at least for Windows; for macOS it should also be possible if you use the 'open -n application.app' script to launch it). Per the above, we'd likely add this as an app descriptor tag so that it's opt-in behaviour:
```<application><allowMultipleInstances>true</allowMultipleInstances></application>```

**Reply by marchbold** [reply]
One thing that has annoyed me with `apm` is the "invocation forwarded to primary instance", which occurs when `adl` is running (i.e. if `apm` is running). Would this work around that issue as well?  This also causes issues in intellij occasionally, particularly if an instance of adl crashes or stalls for some reason. IntelliJ keeps trying to forward to it instead of starting a new instance.

**Reply by ajwfrost** [reply]
For that, you can already add the flag to ADL for this to work for the command-line tool: just update the launch scripts so that it includes the option `-multi-instance` before the `-cmd`. Sorry, I should have highlighted this one for you!

The way it works internally is via local connections I believe, so as an application starts up it will register its app id to make it available for local connections; when an app starts, it looks to see if there's a local connection with its app id that's already present. If so, it forwards the start-up parameters to the original one, and then quits with that "invocation forwarded" message. There is a problem where an app crashes/hangs and the local connection then stays around in the system (presumably it's cleaned up after some timeout) and a comment in the code acknowledges this as a potential issue but without any decent way of solving it..

**Reply by marchbold** [reply]
Ah brilliant!! Didn't know you had done that... I'll test it out first thing tomorrow.

**Reply by marchbold** [reply]
Brilliant!! This makes it work so much better!!! :) Even debugging in intellij doesn't affect it...

**Reply by atollontech** [reply]
It would be very welcome feature to run multiple windowed air apps (win, mac)...

**Reply by marchbold** [reply]
@atollontech Just to clarify you can already develop an app with multiple windows, the issue here is about running multiple instances of the same app :)

**Reply by atollontech** [reply]
We're developing multi-window app (recently integrated Distriqt ANE for webview in popup window). What we need is also multi-app multi-window :) 

Several instances of app running at the same time...

**Reply by chazhenry** [reply]
@ajwfrost  I've upgraded my project structure to v889. Still, on both PC and Mac, when I compile with IntelliJ I get:
`Failed to package AIR application xx.app: /Users/apple/AS3Projects/xx/out/production/TelestratorAIRLoader/TelestratorAIRLoader-app.xml(7): error 103: application.allowMultipleInstances is an unexpected element/attribute`

ADT command line:
`/Applications/IntelliJ IDEA.app/Contents/jbr/Contents/Home/bin/java -Dapplication.home=/Users/apple/Documents/HarmanSDK-Mac/33.1.1.889 -Dfile.encoding=UTF-8 -Djava.awt.headless=true -Duser.language=en -Duser.region=en -Xmx512m -jar /Users/apple/Documents/HarmanSDK-Mac/33.1.1.889/lib/adt.jar -package -storetype PKCS12 -keystore /Applications/IntelliJ IDEA.app/Contents/plugins/flex/lib/temp_keystore.p12 -storepass keystore_password -tsa none -target bundle /Users/apple/AS3Projects/xx/out/production/xx/ChalkTalkTelestrator.app /Users/apple/AS3Projects/xx/out/production/xx/TelestratorAIRLoader-app.xml -C /Users/apple/AS3Projects/xx/out/production/xx/TelestratorAIRLoader.swf -C /Users/apple/AS3Projects/xxicons -C /Users/apple/AS3Projects/xx/lib splash_image.png -C /Users/apple/AS3Projects/xx/src app.lego -C /Users/apple/AS3Projects/xx/sounds`

**Reply by ajwfrost** [reply]
Hi @chazhenry  - sorry, I should have been clearer in terms of the release plans .. the .889 isn't considered quite so "major"! we are putting together a new version, 50.0.x, which will include a lot of new functionality such as this change. So we'll be pushing this out as a pre-release version soon..

thanks

**Reply by 2jfw** [reply]
> [...] ]putting together a new version, 50.0.x, which will include a lot of new functionality such as this change.  

Mm, slightly irritated by the version name **50** being neither _SWF version_ nor _SDK version_ .. 🤔 or are you making a jump from SWF version 44 to 50?

**Reply by ajwfrost** [reply]
:-)   yes we thought it would be nice to have the SWF version and the namespace version actually match!

Part of the reason here is that SWF version 44 (and 45?) and the version naming of 33 (and 34) are already in use, for Flash Player in China that's currently supported by another of Adobe's partners. So there's a bit of a risk of confusion. We'd already updated our numbering for Flash Player (outside of China) to v50, try to try help folk distinguish between Adobe's/Microsoft's versions and ours, so are now updating AIR to match it....

So we will be publishing an AIR SDK with version 50.0.x.x, and updated ActionScript APIs available to content with SWF version 50, with the app descriptor namespace moving to 50.0. And hopefully we'll then actually (a) keep things more in sync, and (b) work more strictly when it comes to changing anything in the app descriptor format or the APIs (resulting in 50.1, 51.0 etc...)

**Reply by 2jfw** [reply]
Nice - I do much appreciate the unification 👍

**Reply by chazhenry** [reply]
@ajwfrost  Can you give us any previews of the 'lot of new functionality'?

**Reply by ajwfrost** [reply]
Some of the updates are things like the requested support for a Timezone class, display cut-out info for mobiles, plus some updates to ANE functions (e.g. getting window handles, supporting delayed loading), encryption of swfs/embedded data, and the start of some new media capabilities. It will be fun writing the release notes .. and enough AS3 API changes that we are finally going to get round to hosting our own version of the AS3 online documentation :-)
